### DISCUSSION TITLE
FSL/Matlab

### DISCUSSION THREAD

**Discussion by jpeelle** [automaticanalysis/automaticanalysis#230] [Ideas]
No matter what I do to set my shell path, Matlab/aa can't find it. I add this to my user script:

```
FSLbinaryDirectory = '/usr/local/fsl/bin'; 
currentPath = getenv('PATH');
if isempty(strfind(currentPath,FSLbinaryDirectory))
    correctedPath = [ currentPath ':' FSLbinaryDirectory ];
    setenv('PATH', correctedPath);
end
```

which works. Presumably this would also work in `startup.m`. I've looked through the FSL documentation and don't see this (but we/aa may call FSL differently than they expect). My question is for others - have you run into this issue? I have been trying to comment/edit the demo script so that it works for me and documenting along the way things I had to change (that I think would not be obvious so some users).

**Comment by jooh** [comment]
Have you tried setting this in AA instead, using

https://github.com/automaticanalysis/automaticanalysis/blob/ee48221924755790fc91d4170ce9da6fc8d566df/aa_parametersets/aap_parameters_defaults.xml#L38-L40

One thing to look out for is changes to your shell environment. If you are normally on bash shell you can see that the above lines actually tries to use csh for the FSL commands. It's easy to have something on your path in one shell but not the other.

**Comment by tiborauer** [comment]
To add to @jooh's comment, `fslsetup` and `fslshell` should correspond.

And here are how setup scripts should look like for various shells: https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FslInstallation/ShellSetup

**Comment by jpeelle** [comment]
I do my best to use bash for everything. From my user xml file:

```
<fslshell desc='Shell used to run FSL' ui='text'>bash</fslshell>
<fsldir desc='Path to fsl' ui='dir'>/usr/local/fsl</fsldir>                  
<fslsetup desc='Path to fsl setup script, executing before any fsl command' ui='text'></fslsetup>
```

and I think I have my profile set up correctly:

```
3145-AD-00084:~ peelle$ echo $FSLDIR
/usr/local/fsl
3145-AD-00084:~ peelle$ which fsl
/usr/local/fsl/bin/fsl
```
but without adding those lines to the matlab code I could not access fsl from within Matlab (and thus, not within aa). After running the user script with all of the above, and working in my shell, here is Matlab:

```
>> getenv('PATH')

ans =

    '/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/bin'
```

so I have been using setenv to get around this. Any other thoughts?

**Comment by tiborauer** [comment]
You do not need to setup your profile, but you need an fsl setup script you specify in `fslsetup`.

**Comment by jpeelle** [comment]
It's not clear to me what goes in the fsl setup script. What kind of file is this? I'm sorry if I'm missing something obvious. I don't see this mentioned on the FSL wiki (I just see instructions for setting up the shell) and I can run FSL via the GUI just fine.

**Comment by jpeelle** [comment]
I tried looking in the aa documentation also but the website is still down...

**Comment by tiborauer** [comment]
The same commands you used to set up the profile should go into the script, so it will be executed. It is a simple text file.

**Comment by jooh** [comment]
It's the same script that's referenced in the installation link Tibor provided:

${FSLDIR}/etc/fslconf/fsl.sh

(actually, with our new fancy environment variable expansion system you might be able to just set it exactly like that in AA and have it expanded to the full path?)

**Comment by tiborauer** [comment]
Actually, it only works if FSLDIR is already set up. Originally we do not assume any profile set up, so if the script contains everything as on the [FSLWiki](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FslInstallation/ShellSetup):
```
FSLDIR=/usr/local/fsl
. ${FSLDIR}/etc/fslconf/fsl.sh
PATH=${FSLDIR}/bin:${PATH}
export FSLDIR PATH
```

then it should work without any profile setting.

**Comment by rhodricusack** [comment]
I have often had difficulty with this - despite some attention I've found
it difficult to work out exactly which login scripts (.bashrc .profile
.login etc) get run when a non-interactive shell is started. I've found the
fail safe to be to add the path as a prefix to each fsl command using
fslsetup.

**Comment by tiborauer** [comment]
I wonder whether we should rather adjust to the user's shell rather than using the specific shells for tools. We could combine the `...shell` parameters into a single `usershell` parameter, which could be even automatically detected if not specified by the user.

**Reply by jones-michael-s** [reply]
Is it ever that case that a tool requires a specific shell? For example, the current aa setup would allow FSL to run in, say, bash, and Freesurfer to run in csh. 

I'm not saying this example is true, but we might think about it before reducing flexibility.

**Reply by tiborauer** [reply]
I have not used many shells but all tools worked in BASH for me, so I guess most tools work in most shells. It is possible that they switch to a different shell, but it does not matter from the user's perspective.
